File: day01/pow.js
function powlogn(x, n) {
  if (n === 0) {
    return 1
  }

  let t = powlogn(x, Math.floor(n / 2))
  if (n % 2) {
    return x * t * t
  }
  return t * t
}

function pown(x, n) {
  let sum = 1
  for (let i = 0; i < n; i++) {
    sum = sum * x
  }
  return sum
}

function f(n) {
  if (n === 0) return 1
  return f(n - 1) + f(n - 1)
}

for (let i = 1; i <= 20; i++) {
  console.time(`2^${i}, nlogn, time = `)
  let result = powlogn(2, i)
  console.timeEnd(`2^${i}, nlogn, time = `)

  console.time(`2^${i}, n, time = `)
  result = pown(2, i)
  console.timeEnd(`2^${i}, n, time = `)

  console.time(`2^${i}, Math.pow, time = `)
  result = Math.pow(2, i)
  console.timeEnd(`2^${i}, Math.pow, time = `)

  console.time(`${i}, f, time = `)
  result = f(i)
  console.timeEnd(`${i}, f, time = `)
}

File: day01/pow.cpp
#include <iostream>
#include <ctime>
#include <cmath>

using namespace std;

double pownlogn(double x, int n) {
    if (n == 0) {
        return 1.0;
    }
    double t = pownlogn(x, n / 2);
    if (n % 2) {
        return x * t * t;
    }
    return t * t;
}

double pown(double x, int n) {
    double sum = 1.0;
    for (int i = 0; i < n; i++) {
        sum = sum * x;
    }
    return sum;
}

int f(int n) {
    if (n == 0) return 1;
    return f(n - 1) + f(n - 1);
}

int main() {

    for (int i = 0; i <= 200; i++) {
        clock_t startTime = clock();
        int result = pownlogn(2, i);
        clock_t endTime = clock();
        cout << "2^" << i << "=" << result << ", nlogn, time = " << double(endTime - startTime) / CLOCKS_PER_SEC << endl;

        startTime = clock();
        result = pown(2, i);
        endTime = clock();
        cout << "2^" << i << "=" << result << ", n, time = " << double(endTime - startTime) / CLOCKS_PER_SEC << endl;

        startTime = clock();
        result = pow(2, i);
        endTime = clock();
        cout << "2^" << i << "=" << result << ", cmath.pow, time = " << double(endTime - startTime) / CLOCKS_PER_SEC << endl;

        // startTime = clock();
        // result = f(i);
        // endTime = clock();
        // cout << i << result << ", f, time = " << double(endTime - startTime) / CLOCKS_PER_SEC << endl;
    }

    return 0;
}